docs/DeveloperGuide.md

```diff
@@ -1,34 +1,216 @@
+- ``SubjectList#addSubject()`` - Adds a new subject to the subject list.
+- ``SubjectList#removeSubject()`` - Removes an existing subject from the subject list.
+- ``SubjectList#listSubjects()`` - List the subjects in the subject list.
 
-## User Stories
+Step 1. Before the user decides to add a flashcard, he/she can create a subject to store the flashcard using
+the command ``addsubject s/SUBJECTNAME``.
+The following diagram describes how the add subject operation works:
+
+![](images/addsubject_sequence_uml.jpg)
+
+Step 2. The user executes the command ``listsubjects`` to view the subjects currently stored in the application.
+
+Step 3. Once the user has chosen a subject, he/she can execute the command ``addcard s/SUBJECTINDEX q/QUESTION a/ANSWER``
+to add a flashcard into the subject. 
+The following diagram describes how the add card operation works:
+
+![](images/addcard_sequence_uml.jpg)
+
+#### 2.1.2. Design Considerations
+##### Aspect: How user can add a flashcard into a subject
+- **Alternative 1 (current choice)**: Include the subject index in the command.
+  - Pros: 
+    - Simple implementation without involving states
+  - Cons: 
+    - The user will have to list the subjects first to determine the subject index.
+  
+- **Alternative 2**: Select the subject first, then add a flashcard.
+  - Pros:
+    - The command for adding a card will be shorter.
+  - Cons:
+    - Multiple states will be involved.
+    - The application will become more complex as different states use different commands.
+    - The application can also become more difficult to use as users can be unclear about the states.
+
+### 2.x. Quiz Feature
+#### 2.x.1. Implementation
+The quiz feature now incorporates random testing, builds upon the subject feature to allows users to set how many questions to quiz for a selected subject. 
+This helps users to quiz by subject, and get a score at the end of each quiz. It implements the following operations:
+
+- ``Quiz#quizQuestion()`` - Outputs a random question to the user that has not been tested before in this quiz session.
+- ``Quiz#quizNext()`` - Retrieves a random question from the available questions for that subject.
+- ``Quiz#markCorrectness()`` - Marks user answer as correct or wrong based on user's judgement.
+
+Given below is an example usage scenario and how the quiz mechanism behaves at each step.
+
+Step 1. Before the user decides to start a quiz, he/she has to create a subject using the command
+ ``addsubject s/SUBJECTNAME``, and add flashcards to that subject using the command ``addcard s/SUBJECTINDEX q/QUESTION a/ANSWER``.
+
+Step 2. The user can start a quiz by indicating a subject and the number of questions to quiz using the command
+``quiz s/SUBJECTINDEX n/NUMBERTOQUIZ``. If the number of questions to quiz is not specified, all the questions stored
+in that subject will be quizzed.
+The following diagram describes how the quiz operation works:
+
+![](images/quiz_sequence_uml.png)
```

Comment:
Is there supposed to be a return arrow at the end?

![image](https://user-images.githubusercontent.com/59989652/77993965-c7d40300-735b-11ea-9bc2-21b92383c315.png)